dvc/repo/experiments/show.py

```diff
@@ -40,85 +39,113 @@ class ExpStatus(Enum):
     Failed = 3
 
 
+class _CachedError(DvcException):
+    def __init__(self, msg, typ, *args):
+        super().__init__(msg, *args)
+        self.typ = typ
+
+
 def _is_scm_error(collected_exp: Dict[str, Any]) -> bool:
     if "error" in collected_exp and (
-        isinstance(collected_exp["error"], (SCMError, InnerScmError))
+        isinstance(collected_exp["error"], (_CachedError, SCMError, InnerScmError))
     ):
         return True
     return False
 
 
-def _show_onerror_collect(result: Dict, exception: Exception, *args, **kwargs):
-    onerror_collect(result, exception, *args, **kwargs)
-    result["data"] = {}
+def _format_exp(exp: SerializableExp) -> Dict[str, Any]:
+    # SerializableExp always includes error but we need to strip it from show
+    # output when it is false-y to maintain compatibility w/tools that consume
+    # json output and assume that "error" key presence means there was an error
+    exp_dict = exp.dumpd()
+    if "error" in exp_dict and not exp_dict["error"]:
+        del exp_dict["error"]
+    return {"data": exp_dict}
+
+
+def _format_error(error: SerializableError):
+    msg = error.msg or "None"
+    return {"data": {}, "error": _CachedError(msg, error.type)}
 
 
-@error_handler
 def collect_experiment_commit(
     repo: "Repo",
     exp_rev: str,
     status: ExpStatus = ExpStatus.Success,
-    param_deps=False,
-    running: Optional[Dict[str, Any]] = None,
-    onerror: Optional[Callable] = None,
+    param_deps: bool = False,
+    force: bool = False,
+    **kwargs,
 ) -> Dict[str, Any]:
-    from dvc.dependency import ParamsDependency, RepoDependency
-
-    result: Dict[str, Any] = defaultdict(dict)
+    cache = repo.experiments.cache
+    # TODO: support filtering serialized exp when param_deps is set
+    if exp_rev != "workspace" and not (force or param_deps):
+        cached_exp = cache.get(exp_rev)
+        if cached_exp:
+            if status == ExpStatus.Running:
+                # expire cached queued exp entry once we start running it
+                cache.delete(exp_rev)
+            elif isinstance(cached_exp, SerializableError):
+                return _format_error(cached_exp)
+            else:
+                return _format_exp(cached_exp)
+    try:
+        exp = _collect_from_repo(
+            repo,
+            exp_rev,
+            status=status,
+            param_deps=param_deps,
+            force=force,
+            **kwargs,
+        )
+        if exp_rev != "workspace" and not param_deps:
+            cache.put(exp, force=True)
+        return _format_exp(exp)
+    except Exception as exc:  # noqa: BLE001, pylint: disable=broad-except
```

Comment:
I don't think we should save cache if there are any errors in it. We don't know if it was just a transient network error like `FileNotFoundError`/`RemoteDepsMissingError`, etc or a corrupted dvc.yaml file.

Reply:
Given that we are only caching experiments by git commit (and not uncommitted workspace states), for things like corrupted dvc.yaml, if we get the error once, we know we will get the error every time (since the git committed dvc.yaml is never going to change).

For things like dvc-tracked metrics files, I suppose there is a possibility that they could be corrupted by a bad `dvc fetch`, in which case the cached exp would contain the error in the nested metrics, but in this case the user can resolve it with something like
```
git checkout <affected commit>
dvc fetch
dvc exp show -force
```

The main point of this PR is to cache fixed git commit states, and if we are not caching commits that contain invalid dvc.yaml files, we will try (and fail) to recollect that commit every time. This is made even worse when users run something like `exp show -A`, given that the affected dvc.yaml file probably exists in a range of commits that all need to be re-collected each time.

Reply:
For reference, the initial version of this PR did not cache entries when we hit errors, but even testing with the vscode demo repo it becomes apparent that if we aren't caching errors, there was almost no overall performance gain due to problematic git commits vastly outnumbering valid ones.

Reply:
Users should not care about caching.

Reply:
If we think this is important, then we should actually improve error handling on repo + metrics/params collection, and differentiate between errors that we know are permanent (a git-committed file is invalid and will never parse) and errors that should be retry-able (a network error occurs trying to fetch something during collection), and then only cache certain cases appropriately